core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -614,18 +614,24 @@ private[spark] class ExecutorAllocationManager(
             stageAttemptToNumRunningTask -= stageAttempt
           }
         }
-        // If the task failed, we expect it to be resubmitted later. To ensure we have
-        // enough resources to run the resubmitted task, we need to mark the scheduler
-        // as backlogged again if it's not already marked as such (SPARK-8366)
-        if (taskEnd.reason != Success) {
-          if (totalPendingTasks() == 0) {
-            allocationManager.onSchedulerBacklogged()
-          }
-          if (taskEnd.taskInfo.speculative) {
-            stageAttemptToSpeculativeTaskIndices.get(stageAttempt).foreach {_.remove(taskIndex)}
-          } else {
-            stageAttemptToTaskIndices.get(stageAttempt).foreach {_.remove(taskIndex)}
-          }
+
+        if (taskEnd.taskInfo.speculative) {
```

Comment:
thanks for fixing this, we made a mistake previously.

Reply:
I think `stageAttemptToSpeculativeTaskIndices` also includes successful finished speculative tasks? Remove a successful speculative task will count it into pending speculative tasks again.

Reply:
Note we already remove pending task `stageAttemptToNumSpeculativeTasks(stageAttempt) -= 1`. @Ngone51